### ISSUE TITLE
Pex bypass/avoid using pep517

### ISSUE THREAD

**Issue by avadhanij6si** [pantsbuild/pex#1560]
We are seeing a set of issues crop up over the week because setuptools over v60.0 introduced breaking changes with regards to the version of distutils used. Setting the env variable `SETUPTOOLS_USE_DISTUTILS=stdlib` seems to be a workaround, but I would like to know if there is a documented way to avoid using pep517 via pex? 

When pex is invoked with a constraints file, or pinned requirement of an older setuptools, it's respected all the way until a dependent package that does not fully conform to pep517 comes along, i.e., it has a `pyproject.toml` file, but no `build-system` section and `build-backend` entry. 

When it does come across such a package, pip is invoked with the following arguments - `'setuptools>=40.8.0' wheel
`. This is causing the latest setuptools to get pulled and installed.

Please excuse any mistaken assumptions I may have made along the way, and I would really appreciate any direction on this matter.

**Comment by jsirois** [comment]
What version of Pex are you using and do you have any compact examples I can exercise using the Pex CLI to see what you're talking about more clearly?

**Comment by avadhanij6si** [comment]
I have tried this with both 2.1.16, and 2.1.56, and I see the same result. To reproduce the issue, you can try any of the following variants - 

```
pex setuptools==44.1.0 wheel==0.33.5 sqlalchemy==1.3.15 -vvv -o testpex --disable-cache
pex setuptools==44.1.0 wheel==0.33.5 sqlalchemy==1.3.15 -vvv -o testpex --disable-cache --constraints constraints.txt
pex sqlalchemy==1.3.15 -vvv -o testpex --disable-cache --constraints constraints.txt
```

The contents of `constraints.txt` - 
```
setuptools==44.1.0
wheel==0.33.5
```

If you inspect the logs, you should see the following in a sequence in the logs

1. `Successfully installed setuptools-60.0.5 wheel-0.37.1`
2. `Successfully built SQLAlchemy`
3. `Successfully installed wheel-0.33.5`
4. `Successfully installed setuptools-44.1.0`
5. `Successfully installed SQLAlchemy-1.3.15`

The first line seems to be happening because I can see `pex` issue the following command through `pip` - 

` Installing build dependencies: started
  Running command /usr/local/Homebrew/Cellar/python@3.7/3.7.12_1/Frameworks/Python.framework/Versions/3.7/bin/python3.7 /private/var/folders/nf/m0dthsxn7gx2v7yzzz6sz8br0000gp/T/tmpdj3pl0an/pip.pex/aef609891d42d65c887d1aeee58c46f6713a7e49/.deps/pip/pip install --ignore-installed --no-user --prefix /private/var/folders/nf/m0dthsxn7gx2v7yzzz6sz8br0000gp/T/pip-build-env-7ofvfqpe/overlay --no-warn-script-location -v --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.8.0' wheel`
  
When I go through the `pyproject.toml` of sqlalchemy==1.3.15, it has no section for `build-system`, and it seems like this is the default behavior.

**Comment by jsirois** [comment]
Ok. I can play with this more tomorrow, but you might try the undocumented Pex feature of `-r requirements.txt` being a way to pass pip options for individual requirements.

**Comment by jsirois** [comment]
I cannot reproduce. Do you have any other details about your environment that might be relevant? Clearly you're on a Mac which I don't have access to.

This is what I find:
[log.txt](https://github.com/pantsbuild/pex/files/7775039/log.txt)

N.B.: Setuptools is now up to 60.1.0 as of ~16 hours ago. Maybe they fixed something here?

**Comment by avadhanij6si** [comment]
So I just checked that log file, and I can see this line - `Successfully installed setuptools-60.1.0 wheel-0.37.1` (line 2960), and the one causing it is the `pip install` command issued at line 1598?

My question really is, why is Pex downloading them, even when I supply my own setuptools==44.1.0 and wheel==0.33.5 requirement? I can see in the sequence that Pex eventually uninstalls those latest wheels, and then installs the ones I am asking it to install. 

Is there a way to tell Pex to use the setuptools and wheel I am supplying it? The reason I mention pep517, is because this behavior seems to be driven by that requirement - https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/?highlight=isolated%20environment#fallback-behaviour

Sqlalchemy 1.3.15 in this case has a pyproject.toml file that does not specify a `build-system` section.

**Comment by jsirois** [comment]
Unfortunately there is no Pex there. Pex calls Pip. It vendors 20.3.4 to be precise. If you can play with the Pip CLI at that version and get things working like you want, then we can translate that into a Pex CLI.

**Comment by jsirois** [comment]
My hint before about requirements files as the means to pass the requirements is due to this: https://pip.pypa.io/en/stable/reference/requirements-file-format/#supported-options

You can pass some Pip flags directly to Pip that way to influence behavior.

**Comment by avadhanij6si** [comment]
If my understanding is correct, the two pip options I would like to use in this situation are - 

```
--no-build-isolation        Disable isolation when building a modern source distribution. Build dependencies specified by PEP 518 must be already installed if this option isused.
--use-pep517                Use PEP 517 for building source distributions (use --no-use-pep517 to force legacy behaviour).
```

These however are not supported in the requirements file.

> Pex calls Pip

This is the crux, I feel. If I can tell pex to not invoke pip with the options `--ignore-installed` when creating a new environment, or if I can tell pex to first install the setuptools version I am specifically asking it to, that would be clean.

Or do you think perhaps this is a situation where we take advantage of venvs feature of pex? I have never used the feature, but if we have a fresh venv with the setuptools==44.1.0 installed as first step, then pex won't bother with issuing the command at line 1598 in that logfile?

**Comment by jsirois** [comment]
If there are pip options you can confirm work but are not exposed by Pex, I'd be happy to accept a feature request (or implementation!) to add those.

As far as venvs go, that's only a runtime feature for a built PEX file. Your issue is at build time.

**Comment by jsirois** [comment]
@avadhanij6si I'm also not sure how generic your situation is. If its particular, you can always build the problematic distribution(s) out of band and then ask Pex to use these pre-builts. Basically:
```
$ pip wheel --wheel-dir ./wheelhouse <options and requirements>
$ pex -f ./wheelhouse <options and requirements>
```
This will still use PyPI to satisfy resolves, but it will also look in the ./wheelhouse directory and prefer any pre-built wheels there over sdists available on PyPI.

**Comment by avadhanij6si** [comment]
I tried the above workflow, but it was still the same behavior - 

```
$ pip wheel --wheel-dir ./localpypi setuptools==44.1.0
Collecting setuptools==44.1.0
  Using cached setuptools-44.1.0-py2.py3-none-any.whl (583 kB)
Saved ./localpypi/setuptools-44.1.0-py2.py3-none-any.whl
WARNING: You are using pip version 20.3.4; however, version 21.3.1 is available.
You should consider upgrading via the '/Users/xxxxxx/.virtualenvs/py3.7/bin/python3.7 -m pip install --upgrade pip' command.

$ pip wheel --wheel-dir ./localpypi wheel==0.33.5
Collecting wheel==0.33.5
  Using cached wheel-0.33.5-py2.py3-none-any.whl (21 kB)
Saved ./localpypi/wheel-0.33.5-py2.py3-none-any.whl
WARNING: You are using pip version 20.3.4; however, version 21.3.1 is available.
You should consider upgrading via the '/Users/xxxxxx/.virtualenvs/py3.7/bin/python3.7 -m pip install --upgrade pip' command.

$ ls ./localpypi
setuptools-44.1.0-py2.py3-none-any.whl wheel-0.33.5-py2.py3-none-any.whl
```

But after running - `pex setuptools==44.1.0 wheel==0.33.5 sqlalchemy==1.3.15 -f ./localpypi -vvv -o testpex --disable-cache 2>&1 | tee out.txt`, I could still see the following line in the logs - `Successfully installed setuptools-60.1.0 wheel-0.37.1`.

I tried this with both 2.1.16 and 2.1.56.

> If there are pip options you can confirm work but are not exposed by Pex

I am relatively new to `pex`, and so I am trying to understand the flow of Pex. What I am unable to tell is why Pex is launching that command - 
``` Created temporary directory: /private/var/folders/nf/m0dthsxn7gx2v7yzzz6sz8br0000gp/T/pip-build-env-i6jwfooi
  Installing build dependencies: started
  Running command /private/var/folders/nf/m0dthsxn7gx2v7yzzz6sz8br0000gp/T/tmpui59xgz6/venvs/f4fe22a38ad1716580d169deb527663c98f3b86d/0834e1f9bd0cd7db7aaa4f49ec2d93f9d1524586/bin/python3.7 /private/var/folders/nf/m0dthsxn7gx2v7yzzz6sz8br0000gp/T/tmpui59xgz6/venvs/f4fe22a38ad1716580d169deb527663c98f3b86d/0834e1f9bd0cd7db7aaa4f49ec2d93f9d1524586/lib/python3.7/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/nf/m0dthsxn7gx2v7yzzz6sz8br0000gp/T/pip-build-env-i6jwfooi/overlay --no-warn-script-location -v --no-binary :none: --only-binary :none: -i https://pypi.org/simple --find-links ./localpypi -- 'setuptools>=40.8.0' wheel
```

From a highlevel browse of the source code, it seems to be centered around - 

1. https://github.com/pantsbuild/pex/blob/v2.1.16/pex/vendor/_vendored/pip/pip/_internal/pyproject.py#L194
2. https://github.com/pantsbuild/pex/blob/v2.1.16/pex/vendor/_vendored/pip/pip/_internal/build_env.py#L174.

The function in `pyproject.py` - `load_pyproject_toml`, seems to accept an argument called - `use_pep517`, but I don't know how to indicate to pex to set that option when it's running.

**Comment by jsirois** [comment]
@avadhanij6si I should have been more specific, you want:
```
$ pip wheel --no-deps --wheel-dir ./localpypi sqlalchemy==1.3.15 <other options needed to get sqlalchemy to build, like maybe --no-use-pep517>
$ pex setuptools==44.1.0 wheel==0.33.5 sqlalchemy==1.3.15 -f ./localpypi -vvv -o testpex
```

The whole point of the 1st step is to prebuild a wheel for the problematic sqlalchemy distribution. In the second step you ask Pex to use that hard won result.

**Comment by jsirois** [comment]
> What I am unable to tell is why Pex is launching that command

Pex is not launching that command. Pex is shelling out to pip and pip internally does this. As I said above, Pex could add support for passing through flags like --no-use-pep51, but that support does not exist today. As such, the workaround I present, allows you to use Pip directly with all its flags to build the problematic wheel and then use Pex with that result.

**Comment by avadhanij6si** [comment]
Very interesting, and you are right, this is a pip behavior. I was able to confirm it running the `pip wheel` with full verbosity. 

I have attached the logs for the following three, if you are interested in seeing them - 
1. `pip wheel --no-deps --wheel-dir ./wheelhouse sqlalchemy==1.3.15 -vvv --no-cache-dir 2>&1 | tee pipout.txt`
2. `pip wheel --no-deps --no-use-pep517 --wheel-dir ./wheelhouse sqlalchemy==1.3.15 -vvv --no-cache-dir 2>&1 | tee pipout2.txt`
3. `pip wheel --no-deps --no-build-isolation --wheel-dir ./wheelhouse sqlalchemy==1.3.15 -vvv --no-cache-dir 2>&1 | tee pipout3.txt`

The first one clearly follows the behavior currently seen with pex, i.e, installing latest setuptools and wheel. I am not entirely sure of the difference between the latter two, but I think `--no-build-isolation` is the proper option that's future proof. The expectation is that the build environment already has all the tools required, so `pip` doesn't have to issue a `pip install --ignore-installed setuptools>=40.8.0 wheel` command.

The current solution of building the problematic wheels separately and storing them in some cache/server should clearly work, but it does introduce issues when packages are updated, since the cache would need then updates. I think it would be great to expose the `--no-build-isolation` option via pex as well. 

What do you think?
[pipout.txt](https://github.com/pantsbuild/pex/files/7810826/pipout.txt)
[pipout2.txt](https://github.com/pantsbuild/pex/files/7810827/pipout2.txt)
[pipout3.txt](https://github.com/pantsbuild/pex/files/7810829/pipout3.txt)

**Comment by jsirois** [comment]
I agree exposing Pip options is probably the way to go. See my comment here: https://github.com/pantsbuild/pex/issues/1560#issuecomment-1000938621. Thanks for confirming building the wheel this way worked for you, meaning, when the option is exposed in Pex, Pex should work for you without an intermediate wheel build step. I'll label this as a feature request.

I'm not sure when I'll have time to get to this / what release it will be in, but you can track progress here or even ask questions if you decide to implement this yourself with a PR.

**Comment by avadhanij6si** [comment]
I appreciate all the debugging help. If it's not clear, I can rename the issue, but I guess the exact enhancement at this point is to expose `--no-build-isolation` through pex.

**Comment by jsirois** [comment]
@avadhanij6si `--no-build-isolation` is now available in Pex 2.1.62. Just pre-install build dependencies in a venv, install pex in that venv, then run `pex --no-build-isolation ...` and any sdist builds will have access to everything in the venv to perform the build with.

**Comment by avadhanij6si** [comment]
Awesome, that was super quick! To be honest, I was hoping our team would just update their packages to newer versions that fully follow pep518, but that might take a while, so this should be useful.
